File: bindings/typescript/packages/client/src/fwd.ts
// SPDX-License-Identifier: Apache-2.0
// SPDX-FileCopyrightText: Copyright 2026 avatarsd LLC

/**
 * @brief FWD (0x0F) / FIELD (0x10) builders + decoder for the libtracer client
 * SDK (#56, RFC-0004, ADR-0034/0035).
 *
 * RFC-0004 §3 pins the remote-operation envelope the conservative ADR-0034
 * slice deferred: a self-describing FWD frame carries a
 * `read`/`write`/`await`/`subscribe` to a path-as-route, and a FWD{REPLY}
 * routes the result back. These builders produce the EXACT bytes the shared
 * conformance vectors pin (`tests/conformance/vectors/v1/fwd/`, `.../field/`)
 * via the cross-validated core codec — nothing here invents wire structure.
 *
 * Child order (RFC-0004 §B): op, dst (PATH), FIELD? selector, src (PATH), then
 *   - REPLY:  kind (VALUE u8), payload?
 *   - WRITE:  payload?
 *   - AWAIT:  await_timeout? (VALUE u64 ns)
 *   - READ :  nothing
 */

import { TYPE, encode, decode } from '@avatarsd-llc/libtracer';
import type { Opt, Tlv } from '@avatarsd-llc/libtracer';
import { opt, pathTlv, nameTlv } from './tlv.js';

/** @brief FWD `op` discriminant (RFC-0004 §B — the first child, a VALUE u8). */
export const FWD_OP = Object.freeze({
  READ: 0,
  WRITE: 1,
  AWAIT: 2,
  REPLY: 3,
} as const);

/** @brief FWD{REPLY} `kind` discriminant (RFC-0004 §D — a VALUE u8 after `src`). */
export const FWD_KIND = Object.freeze({
  RESULT: 0,
  ERROR: 1,
} as const);

/**
 * @brief Registered wire ERROR codes for a `kind=ERROR` reply's
 * `STATUS{ ERROR{ VALUE u16 } }` payload (RFC-0002 §D registry — the u16 LE
 * identity of each built-in `tr::…` error path).
 *
 * Mirrors the C++ reference `error.hpp` registry / `op_resolve.cpp`
 * `error_code()` map.
 */
export const FWD_ERROR = Object.freeze({
  FRAME_TRUNCATED: 0x0001 /**< @brief tr::frame::truncated */,
  FRAME_INVALID: 0x0002 /**< @brief tr::frame::invalid */,
  FRAME_CRC_FAIL: 0x0003 /**< @brief tr::frame::crc_fail */,
  TLV_NESTING_TOO_DEEP: 0x0010 /**< @brief tr::tlv::nesting_too_deep */,
  NOT_FOUND: 0x0020 /**< @brief tr::path::not_found */,
  INVALID_PATH: 0x0021 /**< @brief tr::path::invalid */,
  PATH_IN_USE: 0x0022 /**< @brief tr::path::in_use */,
  TYPE_MISMATCH: 0x0030 /**< @brief tr::schema::type_mismatch */,
  SCHEMA_NOT_FOUND: 0x0031 /**< @brief tr::schema::not_found */,
  BACKPRESSURE: 0x0040 /**< @brief tr::flow::backpressure */,
  TIMEOUT: 0x0041 /**< @brief tr::flow::timeout */,
  ADDRESS_SHIFT_GAP: 0x0042 /**< @brief tr::flow::address_shift_gap */,
  PERMISSION_DENIED: 0x0050 /**< @brief tr::access::denied */,
  TRANSPORT_DOWN: 0x0060 /**< @brief tr::transport::down */,
  VERSION_MISMATCH: 0x0070 /**< @brief tr::version::mismatch */,
} as const);

/**
 * @brief The canonical `tr::<concept>::<error>` namespace path of each
 * registered code (RFC-0002 §A/§D) — the string identity a NAME-form ERROR
 * carries.
 *
 * Mirrors the C++ reference `error.hpp` `err_path()`.
 */
export const FWD_ERROR_PATH: Readonly<Record<number, string>> = Object.freeze({
  [FWD_ERROR.FRAME_TRUNCATED]: 'tr::frame::truncated',
  [FWD_ERROR.FRAME_INVALID]: 'tr::frame::invalid',
  [FWD_ERROR.FRAME_CRC_FAIL]: 'tr::frame::crc_fail',
  [FWD_ERROR.TLV_NESTING_TOO_DEEP]: 'tr::tlv::nesting_too_deep',
  [FWD_ERROR.NOT_FOUND]: 'tr::path::not_found',
  [FWD_ERROR.INVALID_PATH]: 'tr::path::invalid',
  [FWD_ERROR.PATH_IN_USE]: 'tr::path::in_use',
  [FWD_ERROR.TYPE_MISMATCH]: 'tr::schema::type_mismatch',
  [FWD_ERROR.SCHEMA_NOT_FOUND]: 'tr::schema::not_found',
  [FWD_ERROR.BACKPRESSURE]: 'tr::flow::backpressure',
  [FWD_ERROR.TIMEOUT]: 'tr::flow::timeout',
  [FWD_ERROR.ADDRESS_SHIFT_GAP]: 'tr::flow::address_shift_gap',
  [FWD_ERROR.PERMISSION_DENIED]: 'tr::access::denied',
  [FWD_ERROR.TRANSPORT_DOWN]: 'tr::transport::down',
  [FWD_ERROR.VERSION_MISMATCH]: 'tr::version::mismatch',
});

const FWD_ERROR_NAME: Readonly<Record<number, string>> = Object.freeze(
  Object.fromEntries(Object.entries(FWD_ERROR).map(([k, v]) => [v, k])),
);

const FWD_ERROR_CODE_BY_PATH: Readonly<Record<string, number>> = Object.freeze(
  Object.fromEntries(Object.entries(FWD_ERROR_PATH).map(([code, path]) => [path, Number(code)])),
);

/** @brief The human name of a wire ERROR code (or `UNKNOWN(0x..)`). */
export function fwdErrorName(code: number): string {
  return FWD_ERROR_NAME[code] ?? `UNKNOWN(0x${code.toString(16)})`;
}

/** @brief The canonical `tr::…` path of a wire ERROR code (or `null` when unregistered). */
export function fwdErrorPath(code: number): string | null {
  return FWD_ERROR_PATH[code] ?? null;
}

/**
 * @brief The registered u16 code of a `tr::…` error path (RFC-0002 §A/§D), or
 * 0 when the path is not in the frozen registry.
 */
export function fwdErrorCodeForPath(path: string): number {
  return FWD_ERROR_CODE_BY_PATH[path] ?? 0;
}

/* ----------------------------------------------------------- value nodes --- */

/** @brief A 1-byte VALUE TLV node (the op / kind discriminants). */
function valueU8(v: number): Tlv {
  return { type: TYPE.VALUE, opt: opt(), payload: Uint8Array.of(v & 0xff), children: [], trailer: null };
}

/** @brief A little-endian u32 VALUE TLV node (a FIELD `[N]` index). */
function valueU32(v: number): Tlv {
  const p = new Uint8Array(4);
  new DataView(p.buffer).setUint32(0, v >>> 0, true);
  return { type: TYPE.VALUE, opt: opt(), payload: p, children: [], trailer: null };
}

/** @brief A little-endian u64 VALUE TLV node (the AWAIT `await_timeout` ns). */
function valueU64(v: bigint): Tlv {
  const p = new Uint8Array(8);
  new DataView(p.buffer).setBigUint64(0, v, true);
  return { type: TYPE.VALUE, opt: opt(), payload: p, children: [], trailer: null };
}

/* ----------------------------------------------------------- FIELD (0x10) --- */

/** @brief One level of a `:field` selector chain (RFC-0004 §C, one NAME + optional index). */
export interface FieldLevel {
  /** @brief The level's field name (e.g. `"subscribers"`, `"settings"`). */
  readonly name: string;
  /**
   * @brief The index when this level is an element selector (`[N]`).
   *
   * When set, the level is encoded as `ELEMENT` regardless of
   * {@link FieldLevel.mode}.
   */
  readonly index?: number;
  /**
   * @brief `SCALAR` (no brackets, default), `ELEMENT` (`[N]` / append `[]`), or
   * `WILDCARD` (`[*]` — subscriber-path targets only).
   */
  readonly mode?: 'SCALAR' | 'ELEMENT' | 'WILDCARD';
}

/**
 * @brief Parse a `:field` selector string into its levels.
 *
 * Accepts an optional leading `:`; levels are split on `.`; each level may
 * carry a trailing `[N]`, `[]` (append), or `[*]` (wildcard).
 *
 * `":subscribers[]"` → `[{ name: "subscribers", mode: "ELEMENT" }]`;
 * `":subscribers[3]"` → `[{ name: "subscribers", index: 3, mode: "ELEMENT" }]`;
 * `":settings.deadline_ns"` → `[{ name: "settings" }, { name: "deadline_ns" }]`.
 *
 * @param field the selector string (with or without the leading `:`)
 * @returns the parsed levels
 * @throws {RangeError} on an empty selector or a malformed level
 */
export function parseField(field: string): FieldLevel[] {
  const body = field.startsWith(':') ? field.slice(1) : field;
  if (body.length === 0) throw new RangeError('a :field selector must name at least one level');
  return body.split('.').map((part) => {
    const m = /^([^.[\]]+)(?:\[(\*|\d*)\])?$/.exec(part);
    if (!m) throw new RangeError(`malformed :field level ${JSON.stringify(part)}`);
    const name = m[1];
    if (m[2] === undefined) return { name };
    if (m[2] === '*') return { name, mode: 'WILDCARD' as const };
    if (m[2] === '') return { name, mode: 'ELEMENT' as const };
    return { name, index: Number(m[2]), mode: 'ELEMENT' as const };
  });
}

/** @brief Build a FIELD TLV node from its levels (RFC-0004 §C). */
function fieldTlv(levels: FieldLevel[]): Tlv {
  if (levels.length < 1) throw new RangeError('a FIELD selector needs at least one level');
  if (levels.length > 8) throw new RangeError(`a FIELD selector may have at most 8 levels (got ${levels.length})`);
  const children: Tlv[] = [];
  for (const level of levels) {
    children.push(nameTlv(level.name));
    const mode = level.mode ?? (level.index !== undefined ? 'ELEMENT' : 'SCALAR');
    if (mode === 'WILDCARD') {
      children.push(valueU8(2)); // index_mode=WILDCARD
    } else if (mode === 'ELEMENT') {
      if (level.index !== undefined) children.push(valueU32(level.index)); // [N]
      children.push(valueU8(1)); // index_mode=ELEMENT (append "[]" when no index)
    }
    // SCALAR: just the NAME (index_mode default 0).
  }
  return { type: TYPE.FIELD, opt: opt({ pl: true }), payload: new Uint8Array(0), children, trailer: null };
}

/**
 * @brief Build a FIELD TLV (`type=0x10`, PL=1) from a selector string or
 * pre-parsed levels.
 *
 * Vector-pinned: `field-indexed`, `field-nested`, `field-append`.
 *
 * @param field a `:field` string (see {@link parseField}) or its levels
 * @returns the encoded FIELD TLV bytes
 */
export function encodeField(field: string | FieldLevel[]): Uint8Array {
  return encode(fieldTlv(typeof field === 'string' ? parseField(field) : field));
}

/* ------------------------------------------------------------- FWD (0x0F) --- */

/** @brief Request to build a {@link encodeFwd} frame. `op` selects the child layout. */
export interface FwdRequest {
  /** @brief The operation (a {@link FWD_OP} value). */
  readonly op: number;
  /** @brief Forward route: the unresolved destination path segments (the PATH `dst`). */
  readonly dst: string[];
  /** @brief Accumulated return route: the originator's reply endpoint segments (the PATH `src`). */
  readonly src: string[];
  /** @brief Optional `:field` selector (string or levels). */
  readonly field?: string | FieldLevel[];
  /**
   * @brief Optional payload — a complete TLV's encoded bytes (a VALUE to WRITE,
   * a SUBSCRIBER to subscribe, the RESULT/STATUS of a REPLY). Embedded verbatim.
   */
  readonly payload?: Uint8Array;
  /** @brief REPLY only — the reply {@link FWD_KIND}. */
  readonly kind?: number;
  /** @brief AWAIT only — the `await_timeout` in ns (absent ⇒ the responder's 1 s default). */
  readonly awaitTimeoutNs?: bigint;
}

/** @brief Build the FWD child node for an embedded payload (decode the caller's TLV bytes). */
function payloadNode(payload: Uint8Array): Tlv {
  return decode(payload);
}

/**
 * @brief Build a FWD TLV (`type=0x0F`, PL=1) — the remote-operation envelope
 * (RFC-0004 §B).
 *
 * The children are assembled in the normative order for the request's `op`.
 * Vector-pinned: `fwd-read`, `fwd-write-value`, `fwd-await-timeout`,
 * `fwd-write-subscriber-field`, `fwd-reply-result`, `fwd-reply-error`.
 *
 * @param req the operation, routes, and optional selector/payload/kind/timeout
 * @returns the encoded FWD TLV bytes (one complete frame)
 */
export function encodeFwd(req: FwdRequest): Uint8Array {
  const children: Tlv[] = [valueU8(req.op), pathTlv(req.dst)];
  if (req.field !== undefined) {
    children.push(fieldTlv(typeof req.field === 'string' ? parseField(req.field) : req.field));
  }
  children.push(pathTlv(req.src));

  if (req.op === FWD_OP.REPLY) {
    children.push(valueU8(req.kind ?? FWD_KIND.RESULT));
    if (req.payload !== undefined) children.push(payloadNode(req.payload));
  } else if (req.op === FWD_OP.WRITE) {
    if (req.payload !== undefined) children.push(payloadNode(req.payload));
  } else if (req.op === FWD_OP.AWAIT) {
    if (req.awaitTimeoutNs !== undefined) children.push(valueU64(req.awaitTimeoutNs));
  }

  const fwd: Tlv = {
    type: TYPE.FWD,
    opt: opt({ pl: true }),
    payload: new Uint8Array(0),
    children,
    trailer: null,
  };
  return encode(fwd);
}

/* --------------------------------------------------------------- decoding --- */

/** @brief A decoded FWD frame, its children parsed positionally (RFC-0004 §B order). */
export interface ParsedFwd {
  /** @brief The operation (a {@link FWD_OP} value). */
  readonly op: number;
  /** @brief The `dst` PATH child. */
  readonly dst: Tlv;
  /** @brief The optional FIELD selector child (or `null`). */
  readonly field: Tlv | null;
  /** @brief The `src` PATH child. */
  readonly src: Tlv;
  /** @brief The reply {@link FWD_KIND} (REPLY frames only, else `null`). */
  readonly kind: number | null;
  /** @brief The payload TLV child (WRITE payload / REPLY result / RESULT VALUE), or `null`. */
  readonly payload: Tlv | null;
  /** @brief The AWAIT `await_timeout` ns (else `null`). */
  readonly awaitTimeoutNs: bigint | null;
}

const u8 = (t: Tlv): number => (t.payload.length > 0 ? t.payload[0] : 0);

/** @brief Parse an already-decoded FWD {@link Tlv} positionally (RFC-0004 §B). */
export function parseFwdTlv(fwd: Tlv): ParsedFwd {
  if (fwd.type !== TYPE.FWD) throw new RangeError(`not a FWD TLV (type 0x${fwd.type.toString(16)})`);
  const ch = fwd.children;
  let i = 0;
  if (ch[i]?.type !== TYPE.VALUE) throw new RangeError('FWD: missing op VALUE child');
  const op = u8(ch[i++]);
  if (ch[i]?.type !== TYPE.PATH) throw new RangeError('FWD: missing dst PATH child');
  const dst = ch[i++];
  let field: Tlv | null = null;
  if (ch[i]?.type === TYPE.FIELD) field = ch[i++];
  if (ch[i]?.type !== TYPE.PATH) throw new RangeError('FWD: missing src PATH child');
  const src = ch[i++];

  let kind: number | null = null;
  let payload: Tlv | null = null;
  let awaitTimeoutNs: bigint | null = null;
  if (op === FWD_OP.REPLY) {
    if (ch[i]?.type === TYPE.VALUE) kind = u8(ch[i++]);
    if (i < ch.length) payload = ch[i++];
  } else if (op === FWD_OP.WRITE) {
    if (i < ch.length) payload = ch[i++];
  } else if (op === FWD_OP.AWAIT) {
    if (ch[i]?.type === TYPE.VALUE && ch[i].payload.length >= 8) {
      awaitTimeoutNs = new DataView(ch[i].payload.buffer, ch[i].payload.byteOffset, 8).getBigUint64(0, true);
      i++;
    }
  }
  return { op, dst, field, src, kind, payload, awaitTimeoutNs };
}

/**
 * @brief Decode a complete FWD frame's bytes into its parsed children.
 *
 * @param bytes one complete FWD TLV frame
 * @returns the parsed FWD
 * @throws if the bytes do not decode, or are not a FWD
 */
export function decodeFwd(bytes: Uint8Array): ParsedFwd {
  return parseFwdTlv(decode(bytes));
}

/** @brief The ERROR TLV of a `kind=ERROR` reply's STATUS payload, or `null`. */
function replyErrorTlv(reply: ParsedFwd): Tlv | null {
  const status = reply.payload;
  if (status && status.type === TYPE.STATUS && status.children[0]?.type === TYPE.ERROR) {
    return status.children[0];
  }
  return null;
}

/**
 * @brief The registered wire ERROR code of a `kind=ERROR` reply's
 * `STATUS{ ERROR{ VALUE u16 } }` payload per RFC-0002 (or 0 when absent, or
 * when the ERROR carries the string-form NAME identity — see
 * {@link replyErrorPath}).
 */
export function replyErrorCode(reply: ParsedFwd): number {
  const err = replyErrorTlv(reply);
  const id = err?.children[0];
  if (id?.type === TYPE.VALUE && id.payload.length >= 2) {
    return id.payload[0] | (id.payload[1] << 8); // u16 LE registered code
  }
  return 0;
}

/**
 * @brief The string-form `tr::…` identity of a `kind=ERROR` reply's
 * `STATUS{ ERROR{ NAME utf8-path } }` payload per RFC-0002 (or `null` when the
 * reply carries no ERROR, or a registered-code identity — see
 * {@link replyErrorCode}).
 */
export function replyErrorPath(reply: ParsedFwd): string | null {
  const err = replyErrorTlv(reply);
  const id = err?.children[0];
  if (id?.type === TYPE.NAME) return new TextDecoder().decode(id.payload);
  return null;
}

/** @brief Re-export Opt so consumers building raw nodes do not need the core import. */
export type { Opt };

File: bindings/rust/src/fwd.rs
// SPDX-License-Identifier: Apache-2.0
// SPDX-FileCopyrightText: Copyright 2026 avatarsd LLC

/*!
 * @brief FWD remote-operation envelope — FWD TLV (`0x0F`) build/parse (Unit 4).
 *
 * The source-routed remote-operation frame (RFC-0004 §B/§D, ADR-0035). Mirrors
 * `bindings/typescript/packages/client/src/fwd.ts` and the terminus resolver of
 * `core/src/op_resolve.cpp`. Child order (RFC-0004 §B): `op`, `dst` (PATH),
 * `FIELD?` selector, `src` (PATH), then per-op:
 *   - REPLY: `kind` (VALUE u8), optional payload
 *   - WRITE: optional payload
 *   - AWAIT: optional `await_timeout` (VALUE u64 ns)
 *   - READ : nothing
 *
 * Vector-pinned: all nine vectors under `fwd/`.
 */

use alloc::string::String;
use alloc::vec::Vec;

use crate::error_registry::ErrCode;
use crate::field::{field_tlv, parse_field, FieldLevel};
use crate::path::path_to_tlv;
use crate::tlv_builders::{path as build_path, value_u64, value_u8, BuildError};
use crate::{type_code, Opt, Tlv};

/** @brief FWD `op` discriminant (RFC-0004 §B — the first child, a VALUE u8). */
pub mod fwd_op {
    /** @brief Read a vertex's last-known value. */
    pub const READ: u8 = 0;
    /** @brief Write a value / field (subscribe is a `:subscribers[]` field-write). */
    pub const WRITE: u8 = 1;
    /** @brief Await the next write, up to an optional timeout. */
    pub const AWAIT: u8 = 2;
    /** @brief A reply routed back along the accumulated `src`. */
    pub const REPLY: u8 = 3;

    /**
     * @brief The opcode field of the `op` byte — bits 5-0 (RFC-0024 §7.5).
     *
     * Bits 7-6 are FLAGS. A forwarder MUST mask before switching (RFC-0024 §9.3): an
     * unrecognised flag then degrades to the plain opcode instead of an unknown-opcode
     * reject, which is what makes a flag additive at all.
     */
    pub const OPCODE_MASK: u8 = 0x3F;

    /**
     * @brief `op` bit 7 — the bound-path MINT REQUEST (RFC-0024 §7.5).
     *
     * Set by an origin asking each host on the route to answer with its own vertex ref. It
     * costs ZERO request bytes: there is no free TLV `opt` bit, and a dedicated presence
     * child would spend a 4-byte header to express one. The request is a hint, never an
     * obligation — a host that will not mint answers the ordinary reply.
     */
    pub const FLAG_MINT_REQUEST: u8 = 0x80;
}

/** @brief FWD{REPLY} `kind` discriminant (RFC-0004 §D — a VALUE u8 after `src`). */
pub mod fwd_kind {
    /** @brief A successful result (payload is the RESULT value). */
    pub const RESULT: u8 = 0;
    /** @brief A failure (payload is `STATUS{ ERROR }`). */
    pub const ERROR: u8 = 1;
}

/**
 * @brief The registered wire ERROR code of a `tr::…` path, or 0 when unregistered.
 * The FWD error-path table is [`ErrCode`] itself (RFC-0002 §D).
 */
pub fn fwd_error_code_for_path(path: &str) -> u16 {
    ErrCode::from_path(path).map(ErrCode::code).unwrap_or(0)
}

/** @brief The canonical `tr::…` path of a registered wire ERROR code, or `None`. */
pub fn fwd_error_path(code: u16) -> Option<&'static str> {
    ErrCode::from_code(code).map(ErrCode::path)
}

/* -------------------------------------------------------------- FWD request --- */

/** @brief A `:field` selector for a FWD request: a raw string (parsed) or pre-built levels. */
pub enum FieldSel<'a> {
    /** @brief A `:field` selector string (see [`parse_field`]). */
    Str(&'a str),
    /** @brief Pre-parsed selector levels. */
    Levels(Vec<FieldLevel>),
}

/** @brief A request to build a FWD frame. `op` selects the trailing-child layout. */
pub struct FwdRequest<'a> {
    /** @brief The operation (a [`fwd_op`] value). */
    pub op: u8,
    /** @brief Forward route: the unresolved destination path segments (the PATH `dst`). */
    pub dst: &'a [&'a str],
    /** @brief Accumulated return route: the originator's reply-endpoint segments (`src`). */
    pub src: &'a [&'a str],
    /** @brief Optional `:field` selector. */
    pub field: Option<FieldSel<'a>>,
    /**
     * @brief Optional payload node (a VALUE to WRITE, a SUBSCRIBER to subscribe, the
     * RESULT/STATUS of a REPLY). Embedded verbatim.
     */
    pub payload: Option<Tlv>,
    /** @brief REPLY only — the reply [`fwd_kind`] (defaults to RESULT). */
    pub kind: Option<u8>,
    /** @brief AWAIT only — the `await_timeout` in ns (absent ⇒ the responder default). */
    pub await_timeout_ns: Option<u64>,
}

impl<'a> FwdRequest<'a> {
    /** @brief A minimal request: `op`, `dst`, `src`; all optional fields cleared. */
    pub fn new(op: u8, dst: &'a [&'a str], src: &'a [&'a str]) -> FwdRequest<'a> {
        FwdRequest {
            op,
            dst,
            src,
            field: None,
            payload: None,
            kind: None,
            await_timeout_ns: None,
        }
    }
}

fn field_node(sel: &FieldSel) -> Result<Tlv, BuildError> {
    match sel {
        FieldSel::Str(s) => field_tlv(&parse_field(s)?),
        FieldSel::Levels(l) => field_tlv(l),
    }
}

/**
 * @brief Build a FWD TLV node (`type=0x0F`, `PL=1`) — the remote-operation envelope.
 * Children are assembled in the normative order for the request's `op`.
 * Vector-pinned: `fwd-read`, `fwd-write-value`, `fwd-await-timeout`,
 * `fwd-write-subscriber-field`, `fwd-reply-result`, `fwd-reply-error`,
 * `fwd-routed-mount-residual`, `fwd-routed-two-mount`, `fwd-src-accumulated`,
 * `fwd-wildcard-reject`.
 *
 * # Errors
 * Any error from building the `dst`/`src` PATHs, the FIELD selector, or an
 * invalid `op`.
 */
pub fn encode_fwd(req: &FwdRequest) -> Result<Tlv, BuildError> {
    let mut children = alloc::vec![value_u8(req.op), build_path(req.dst)?];
    if let Some(sel) = &req.field {
        children.push(field_node(sel)?);
    }
    children.push(build_path(req.src)?);

    if req.op == fwd_op::REPLY {
        children.push(value_u8(req.kind.unwrap_or(fwd_kind::RESULT)));
        if let Some(p) = &req.payload {
            children.push(p.clone());
        }
    } else if req.op == fwd_op::WRITE {
        if let Some(p) = &req.payload {
            children.push(p.clone());
        }
    } else if req.op == fwd_op::AWAIT {
        if let Some(ns) = req.await_timeout_ns {
            children.push(value_u64(ns));
        }
    }

    Ok(Tlv {
        type_code: type_code::FWD,
        opt: Opt::structured(),
        payload: Vec::new(),
        children,
        trailer: None,
    })
}

/**
 * @brief Convenience: build a FWD frame directly to encoded bytes.
 *
 * # Errors
 * Any error from [`encode_fwd`].
 */
pub fn encode_fwd_bytes(req: &FwdRequest) -> Result<Vec<u8>, BuildError> {
    Ok(crate::encode(&encode_fwd(req)?))
}

/* --------------------------------------------------------------- FWD parse --- */

/** @brief A parsed FWD frame, its children read positionally (RFC-0004 §B order). */
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct ParsedFwd {
    /** @brief The operation (a [`fwd_op`] value). */
    pub op: u8,
    /** @brief The `dst` route child — a `PATH`, or a `PATH_REF` when [`Self::dst_bound`]. */
    pub dst: Tlv,
    /** @brief The optional FIELD selector child. */
    pub field: Option<Tlv>,
    /** @brief The `src` route child — a `PATH`, or a `PATH_REF` on a reply to a bound op. */
    pub src: Tlv,
    /** @brief The reply [`fwd_kind`] (REPLY frames only). */
    pub kind: Option<u8>,
    /** @brief The payload TLV child (WRITE payload / REPLY result), if present. */
    pub payload: Option<Tlv>,
    /** @brief The AWAIT `await_timeout` ns, if present. */
    pub await_timeout_ns: Option<u64>,
    /** @brief `op` bit 7 was set — the origin asked for a bound-path mint (RFC-0024 §7.5). */
    pub mint_request: bool,
    /** @brief `dst` is a `PATH_REF` (`0x14`), not a canonical `PATH` (RFC-0024 §4). */
    pub dst_bound: bool,
}

fn u8_of(t: &Tlv) -> u8 {
    t.payload.first().copied().unwrap_or(0)
}

/**
 * @brief Parse an already-decoded FWD [`Tlv`] positionally (RFC-0004 §B).
 *
 * # Errors
 * [`BuildError::TypeMismatch`] if the TLV is not a FWD or is missing a required
 * `op` / `dst` / `src` child.
 */
pub fn parse_fwd_tlv(fwd: &Tlv) -> Result<ParsedFwd, BuildError> {
    if fwd.type_code != type_code::FWD {
        return Err(BuildError::TypeMismatch);
    }
    let ch = &fwd.children;
    let mut i = 0;
    if ch.get(i).map(|c| c.type_code) != Some(type_code::VALUE) {
        return Err(BuildError::TypeMismatch);
    }
    // The opcode is `op & 0x3F`; bits 7-6 are flags (RFC-0024 §7.5, normative in §9.3).
    // Switching on the raw byte would make a mint-flagged READ an unknown opcode here.
    let op_byte = u8_of(&ch[i]);
    let op = op_byte & fwd_op::OPCODE_MASK;
    let mint_request = op_byte & fwd_op::FLAG_MINT_REQUEST != 0;
    i += 1;
    // Two address forms (RFC-0024 §4): a canonical PATH, or a bound PATH_REF whose body
    // shape the grammar has already checked. What an element MEANS is node-scoped, so a
    // binding core carries one and never interprets it.
    let dst_bound = ch.get(i).map(|c| c.type_code) == Some(type_code::PATH_REF);
    if !dst_bound && ch.get(i).map(|c| c.type_code) != Some(type_code::PATH) {
        return Err(BuildError::TypeMismatch);
    }
    let dst = ch[i].clone();
    i += 1;
    let mut field = None;
    if ch.get(i).map(|c| c.type_code) == Some(type_code::FIELD) {
        field = Some(ch[i].clone());
        i += 1;
    }
    // `src` takes either form too: a reply to a bound operation echoes the request's `dst`.
    let src_kind = ch.get(i).map(|c| c.type_code);
    if src_kind != Some(type_code::PATH) && src_kind != Some(type_code::PATH_REF) {
        return Err(BuildError::TypeMismatch);
    }
    let src = ch[i].clone();
    i += 1;

    let mut kind = None;
    let mut payload = None;
    let mut await_timeout_ns = None;
    if op == fwd_op::REPLY {
        if ch.get(i).map(|c| c.type_code) == Some(type_code::VALUE) {
            kind = Some(u8_of(&ch[i]));
            i += 1;
        }
        if i < ch.len() {
            payload = Some(ch[i].clone());
        }
    } else if op == fwd_op::WRITE {
        if i < ch.len() {
            payload = Some(ch[i].clone());
        }
    } else if op == fwd_op::AWAIT {
        if let Some(c) = ch.get(i) {
            if c.type_code == type_code::VALUE && c.payload.len() >= 8 {
                await_timeout_ns = Some(c.payload_uint());
            }
        }
    }

    Ok(ParsedFwd {
        op,
        dst,
        field,
        src,
        kind,
        payload,
        await_timeout_ns,
        mint_request,
        dst_bound,
    })
}

/**
 * @brief Decode a complete FWD frame's bytes into its parsed children.
 *
 * # Errors
 * [`BuildError::TypeMismatch`] if the frame is not a decodable FWD (a codec
 * fault surfaces as `TypeMismatch`; use [`crate::decode`] directly for the
 * underlying [`crate::Error`]).
 */
pub fn decode_fwd(bytes: &[u8]) -> Result<ParsedFwd, BuildError> {
    let tlv = crate::decode(bytes).map_err(|_| BuildError::TypeMismatch)?;
    parse_fwd_tlv(&tlv)
}

/**
 * @brief The `dst` route of a parsed FWD in string form.
 *
 * # Errors
 * A PATH-decode error.
 */
pub fn fwd_dst_path(fwd: &ParsedFwd) -> Result<String, BuildError> {
    crate::path::tlv_to_path(&fwd.dst)
}

/**
 * @brief The `src` (accumulated return route) of a parsed FWD in string form.
 *
 * # Errors
 * A PATH-decode error.
 */
pub fn fwd_src_path(fwd: &ParsedFwd) -> Result<String, BuildError> {
    crate::path::tlv_to_path(&fwd.src)
}

/* ------------------------------------------------------------- reply errors --- */

/**
 * @brief The ERROR TLV of a `kind=ERROR` reply's `STATUS{ ERROR }` payload, or `None`.
 *
 * The **first ERROR child of the STATUS, at whatever position** — never `children[0]`.
 * reference/05 §`0x09` gives a STATUS "one or more ERROR TLVs and optional DESCRIPTION
 * text" and pins no order over them; RFC-0002 §C pins position only one level down,
 * INSIDE the ERROR ("its first child is the identity"). A reader that demands position 0
 * refuses a documented STATUS shape and answers code 0, which is what it also answers for
 * a STATUS carrying no ERROR at all.
 *
 * Vector-pinned by `fwd/fwd-reply-error-after-description` (the DESCRIPTION written
 * first), which the TypeScript binding pins against the same bytes so the two cores
 * cannot drift apart on it again (#878).
 */
fn reply_error_tlv(reply: &ParsedFwd) -> Option<&Tlv> {
    let status = reply.payload.as_ref()?;
    if status.type_code == type_code::STATUS {
        return status.first_child(type_code::ERROR);
    }
    None
}

/**
 * @brief The registered wire ERROR code of a `kind=ERROR` reply's
 * `STATUS{ ERROR{ VALUE u16 } }` payload (RFC-0002), or 0 when absent or when
 * the ERROR carries the string-form identity (see [`reply_error_path`]).
 */
pub fn reply_error_code(reply: &ParsedFwd) -> u16 {
    if let Some(err) = reply_error_tlv(reply) {
        if let Some(id) = err.children.first() {
            if id.type_code == type_code::VALUE && id.payload.len() >= 2 {
                return id.payload[0] as u16 | ((id.payload[1] as u16) << 8);
            }
        }
    }
    0
}

/**
 * @brief The string-form `tr::…` identity of a `kind=ERROR` reply's
 * `STATUS{ ERROR{ NAME } }` payload (RFC-0002), or `None` when the reply carries
 * no ERROR or a registered-code identity (see [`reply_error_code`]).
 *
 * # Errors
 * [`BuildError::InvalidUtf8`] if the NAME identity is not valid UTF-8.
 */
pub fn reply_error_path(reply: &ParsedFwd) -> Result<Option<String>, BuildError> {
    if let Some(err) = reply_error_tlv(reply) {
        if let Some(id) = err.children.first() {
            if id.type_code == type_code::NAME {
                return Ok(Some(id.payload_str()?.into()));
            }
        }
    }
    Ok(None)
}

/**
 * @brief Convenience: build a `dst`/`src` PATH from a rooted string (for callers that
 * have string paths rather than segment slices).
 *
 * # Errors
 * A path validation error.
 */
pub fn path_tlv_from_str(text: &str) -> Result<Tlv, BuildError> {
    path_to_tlv(text)
}
